### ISSUE TITLE
Interrupts gone wild in kernel > 3.12.y

### ISSUE THREAD

**Issue by CurlyMoo** [raspberrypi/linux#653]
I'm was successfully running kernel 3.12.20 with pilight. When updating to kernel, i notice that the interrupts are not working properly anymore. In the older kernel all interrupts are processed normally and validly. In the newer version, the interrupts just fire making my gpio read thread spike at 99% CPU.

**Comment by popcornmix** [comment]
Does the dt-blob.bin file from https://github.com/tvdzwan/hyperion/issues/136 avoid the problem?

**Comment by CurlyMoo** [comment]
Sadly, it didn't work.

**Comment by popcornmix** [comment]
Can you identify the exact firmware/kernel commit that caused this?

**Comment by CurlyMoo** [comment]
The firmware is not part of the issue. I've successfully tested firmware

```
Aug  1 2014 18:05:07
Copyright (c) 2012 Broadcom
version 158db03216ab9f49243655d4d325da3264567104 (tainted) (release)
```

on a working kernel.

As in the kernel. It's hard to tell what the exact commit was, because i don't have my cross-compile environment available anymore. Maybe @mk01 knows what commits where used in a specific kernel. What i can tell is that the XBian kernel `3.12.20` works fine (released through apt on 26-05-2014) but all kernels from `3.13.7+` fail (released through apt on 25-05-2014).

**Comment by popcornmix** [comment]
Unless the problem occurs with raspbian kernel then I can't help.
The issue may well be caused by patches or config options chosen by xbian that aren't used by raspbian.

If you can reproduce the issue with a raspbian image then using rpi-update you can identify the exact kernel update, and we may be able to do something. If the problem only occurs with xbian then you need to report the issue there.

**Comment by mk01** [comment]
@CurlyMoo 

there are no patches for RPI kernels. 
could it be change in time accounting TICK timing -> IRQ timings ?

**Comment by popcornmix** [comment]
Can you try reverting this: https://github.com/raspberrypi/linux/commit/39e4e12d12c19bb79a4c3a30360faf9b504222dc

**Comment by mk01** [comment]
ok, will rebuild with this reverted

**Comment by mk01** [comment]
@CurlyMoo 

xbian-package-kernel3.15.8+-1407341207.deb in staging

**Comment by CurlyMoo** [comment]
Nope.

**Comment by popcornmix** [comment]
That's the only interrupt related change I can think of in recent history.
Does plight use 1-wire driver? Which gpio pin?
Will really need to bisect to find out which commit caused the problem.

**Comment by CurlyMoo** [comment]
To bisect things i have used the `isr` example of wiringPi. Nothing changed as in the gpio connections or other software. Just the kernel.

Compare the output on the failing kernel (the first 15 results):

```
Waiting ...  Done. counter: 465470
Waiting ...  Done. counter: 716709
Waiting ...  Done. counter: 967771
Waiting ...  Done. counter: 1218740
Waiting ...  Done. counter: 1469698
Waiting ...  Done. counter: 1720582
Waiting ...  Done. counter: 1971454
Waiting ...  Done. counter: 2222343
Waiting ...  Done. counter: 2473338
Waiting ...  Done. counter: 2724468
Waiting ...  Done. counter: 2975497
Waiting ...  Done. counter: 3226439
Waiting ...  Done. counter: 3477889
Waiting ...  Done. counter: 3729305
Waiting ...  Done. counter: 3980640
```

And the working kernel:

```
Waiting ...  Done. counter:     1
Waiting ...  Done. counter:     2
Waiting ...  Done. counter:     3
Waiting ...  Done. counter:     5
Waiting ...  Done. counter:     8
Waiting ...  Done. counter:    10
Waiting ...  Done. counter:    11
Waiting ...  Done. counter:    12
Waiting ...  Done. counter:    13
Waiting ...  Done. counter:    16
Waiting ...  Done. counter:    17
Waiting ...  Done. counter:    18
Waiting ...  Done. counter:    19
Waiting ...  Done. counter:    20
Waiting ...  Done. counter:    21
```

I believe the failing part of the is here https://github.com/WiringPi/WiringPi/blob/master/wiringPi/wiringPi.c#L1014-L1038

**Comment by popcornmix** [comment]
So the problem is only on 3.13 kernel tree?
I thought from the issue title that 3.12 broke sometime after 3.12.20 (raspian is now on 3.12.26).
3.13 is not something we are actively supporting.
3.12 is still the recommended stable kernel (used by raspbian) and it has long term support.

**Comment by CurlyMoo** [comment]
As i posted earlier, the problems first appears in 3.13 but is also present in 3.15.

**Comment by CurlyMoo** [comment]
> 3.13 is not something we are actively supporting.

And never will? The whole idea is that we try to fix issues when they are discovered so the switch can be made faster, right? Also, because the bug is present in both 3.13 and 3.15.

**Comment by popcornmix** [comment]
If it affects 3.13 and 3.15 (and presumably later kernels) then we'd like it understood and fixed.

We offically support the kernel used in raspbian (3.12 currently, but it may be bumped to 3.14 which is also LTS).

I also port to the latest stable kernel and track updates of that (currently 3.15, but that will switch to 3.16 when I've tracked down the upstream bug that's broken the network).

3.13 is an abandoned kernel tree (not just by us, but by https://www.kernel.org/). You should probably stick with the kernel that raspbian uses (which has LTS and recevies regular updates), or if you want to be bleeding edge (like openelec) then picking the latest stable version is another valid choice.

I believe the set of Pi specific patches are the same in 3.12 and 3.15 trees (with minor conflict resolves). I'll have a closer look soon just to confirm that.

If you are able to test using openelec (which is currently using 3.15 kernel) I'd be very interested in the outcome. openelec has been using 3.15 for some time with various gpio related peripherals in use, and I've not seen this problem reported.

**Comment by CurlyMoo** [comment]
Just tested OpenElec 4.0.7 with kernel 3.14.11 on GPIO 18. I just connected the 3.3v directly to GPIO 18 and reconnected it once to trigger at least one interrupt. Same results:

```
Waiting ...  Done. counter: 227244
Waiting ...  Done. counter: 434937
Waiting ...  Done. counter: 642668
Waiting ...  Done. counter: 850459
Waiting ...  Done. counter: 1058213
Waiting ...  Done. counter: 1265647
Waiting ...  Done. counter: 1473434
Waiting ...  Done. counter: 1680847
Waiting ...  Done. counter: 1888743
Waiting ...  Done. counter: 2096497
Waiting ...  Done. counter: 2304342
Waiting ...  Done. counter: 2512152
Waiting ...  Done. counter: 2719996
Waiting ...  Done. counter: 2927740
Waiting ...  Done. counter: 3135484
```

Also tested OpenElec 4.1.4 with kernel 3.16.1

```
Waiting ...  Done. counter: 322447
Waiting ...  Done. counter: 531665
Waiting ...  Done. counter: 740912
Waiting ...  Done. counter: 950106
Waiting ...  Done. counter: 1159553
Waiting ...  Done. counter: 1368713
Waiting ...  Done. counter: 1577942
Waiting ...  Done. counter: 1786891
Waiting ...  Done. counter: 1995860
Waiting ...  Done. counter: 2204718
Waiting ...  Done. counter: 2413523
Waiting ...  Done. counter: 2622183
Waiting ...  Done. counter: 2830875
Waiting ...  Done. counter: 3039274
Waiting ...  Done. counter: 3248302
```

Again tested with the same `ISR` program from wiringPi.

When i manually check the value of `/sys/class/gpio/gpio18/value` it does correctly represent the actual state of the GPIO pin.

**Comment by popcornmix** [comment]
@CurlyMoo 
Can you confirm:
1. when GPIO18 is forced high (e.g. connected to 3.3V through a resistor)
2. when GPIO18 is forced low (e.g. connected to GND through a resistor)

that the interrupt count keeps incrementing.

Can you also see these interrupts through /proc/interrupts?

**Comment by CurlyMoo** [comment]
Still on kernel 3.16.1 latest OpenElec 4.1.4. In each case:
1. i will connect the cable
2. disconnect the cable
3. reconnect the cable
(also in the GND case, just to be sure)

**VCC > 22Ohm > GPIO18**
Terminal 1

```
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
1
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    2969881   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5274793   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       1774   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6206177   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41293   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      24490   ARMCTRL  mmc0
188:       1774      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
0
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    2981486   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5297624   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       1780   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6216529   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41312   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      24561   ARMCTRL  mmc0
188:       1780      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
1
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    2989364   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5313461   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       2082   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6222589   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41324   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      24611   ARMCTRL  mmc0
188:       2082      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
```

Terminal 2

```
Waiting ...  Done. counter: 282676
Waiting ...  Done. counter: 492183
Waiting ...  Done. counter: 700892
Waiting ...  Done. counter: 909566
Waiting ...  Done. counter: 1118246
Waiting ...  Done. counter: 1327711
Waiting ...  Done. counter: 1536424
Waiting ...  Done. counter: 1745076
Waiting ...  Done. counter: 1954591
Waiting ...  Done. counter: 2163528
Waiting ...  Done. counter: 2372648
Waiting ...  Done. counter: 2581791
Waiting ...  Done. counter: 2791608
Waiting ...  Done. counter: 3000778
Waiting ...  Done. counter: 3209877
```

**GND > 22Ohm > GPIO18**

```
Terminal 1
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
0
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    3045654   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5416610   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       2083   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6343603   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41413   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      24935   ARMCTRL  mmc0
188:       2083      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
0
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    3054677   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5433037   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       2083   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6362682   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41430   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      24987   ARMCTRL  mmc0
188:       2083      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
OpenELEC:~ # cat /sys/class/gpio/gpio18/value
0
OpenELEC:~ # cat /proc/interrupts
           CPU0
  3:    3061734   ARMCTRL  BCM2708 Timer Tick
 16:          0   ARMCTRL  bcm2708_fb dma
 32:    5445658   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 52:       2083   ARMCTRL  BCM2708 GPIO catchall handler
 65:          9   ARMCTRL  ARM Mailbox IRQ
 66:    6377383   ARMCTRL  VCHIQ doorbell
 75:          1   ARMCTRL
 77:      41439   ARMCTRL  bcm2708_sdhci (dma)
 79:         14   ARMCTRL  bcm2708_i2c.0, bcm2708_i2c.1
 80:          0   ARMCTRL  bcm2708_spi.0
 84:      25028   ARMCTRL  mmc0
188:       2083      GPIO  gpiolib
FIQ:              usb_fiq
Err:          0
```

Terminal 2 stays nice and quit

```
Waiting ...
```

**Comment by popcornmix** [comment]
Do you see this straight from boot on a clean openelec install, or only after you have enabled pilight?

**Comment by CurlyMoo** [comment]
Procedure.
1. Install OpenElec
2. Enable SSH
3. Mount NFS share
4. Disable lirc
5. Remove `lirc_rpi` and `lirc_dev` kernel modules
6. Start staticly compiled `isr` (and `gpio` program) from wiringPi.

Nothing more.

**Comment by popcornmix** [comment]
Can you upload (or point to) the static wiringpi you are using?
I'd like to be able to reproduce this.

**Comment by CurlyMoo** [comment]
http://www.pilight.org/isr
http://www.pilight.org/gpio

Just put both scripts in the same folder.

**Comment by CurlyMoo** [comment]
Let me know when you downloaded them...

**Comment by popcornmix** [comment]
So what commands do I execute to see bad behaviour?
I have no hardware attached - should I still see spurious interrupts?

**Comment by CurlyMoo** [comment]
Just run `./ISR` and if you connect and disconnect the forced HIGH on GPIO 18, you should see interrupts sparking and the CPU usage go up.

**Comment by popcornmix** [comment]
Connecting and disconnecting will cause multiple interrupts (http://www.labbookpages.co.uk/electronics/debounce.html).
Do you interrupts occurring while the GPIO is securely held low or high?

**Comment by CurlyMoo** [comment]
The interrupts occur even when the GPIO is held high yes (after 1 triggered interrupt).

**Comment by popcornmix** [comment]
At a guess I'd say that the irq executable enables gpio "high" interrupts.
Now the thing with high (or low) interrupts is they continue interrupting unless the high (or low) gpio line goes away.

My guess why this wasn't a problem on the older kernel is we didn't use to support "high" and "low" interrupts. This was added in the 3.10 tree:
https://github.com/raspberrypi/linux/commit/aefeaf21a7612ad300281760b86c3e1984ae0765

I'd need to see the source of "irq", but I suspect that it is requesting a high interrupt, when a rising edge interrupt was actually wanted.

**Comment by CurlyMoo** [comment]
> I'd need to see the source of "irq"

Did you mean wiringPi `isr`?

## 

> My guess why this wasn't a problem on the older kernel is we didn't use to support "high" and "low" interrupts. This was added in the 3.10 tree:

How could the 3.12.y kernel work fine if that's the case?

**Comment by popcornmix** [comment]
Yes - is this what you are using?
https://git.drogon.net/?p=wiringPi;a=snapshot;h=df45388f6431f7baba31ac1e8e242d89828637a0;sf=tgz
That seems to just use falling edge interrupts - have you edited the code?

**Comment by CurlyMoo** [comment]
The only edit i did was changing the pin from 0 to 1 and changing the search path of the gpio executable.

I used the source as found here:
https://github.com/WiringPi/WiringPi/

**Comment by popcornmix** [comment]
That code looks to be over a year old.
Can you try with https://git.drogon.net/?p=wiringPi;a=summary
which has had updates a month ago.

**Comment by CurlyMoo** [comment]
Did you read my additional comment here: https://github.com/raspberrypi/linux/issues/653#issuecomment-52544585

**Comment by CurlyMoo** [comment]
Just tested the latest version of wiringPi (the two online binary's are updated as well) but it still doesn't work:

```
Waiting ...  Int on pin 1: Counter: 224650
Waiting ...  Int on pin 1: Counter: 423949
Waiting ...  Int on pin 1: Counter: 622449
Waiting ...  Int on pin 1: Counter: 820854
Waiting ...  Int on pin 1: Counter: 1019357
Waiting ...  Int on pin 1: Counter: 1217741
Waiting ...  Int on pin 1: Counter: 1416220
Waiting ...  Int on pin 1: Counter: 1614697
Waiting ...  Int on pin 1: Counter: 1813187
Waiting ...  Int on pin 1: Counter: 2011679
```

Again, trigger at least one interrupt for it to go wild (in this case 18).

Or GPIO 17:

```
Waiting ...  Int on pin 0: Counter: 370822
Waiting ...  Int on pin 0: Counter: 566937
Waiting ...  Int on pin 0: Counter: 763099
```

Or GPIO 21/27:

```
Waiting ...  Int on pin 2: Counter: 376948
Waiting ...  Int on pin 2: Counter: 578794
Waiting ...  Int on pin 2: Counter: 779952
```

GPIO 22

```
Waiting ...  Int on pin 3: Counter: 383021
Waiting ...  Int on pin 3: Counter: 574119
Waiting ...  Int on pin 3: Counter: 765268
```

You get the picture.

Top output when running `isr` with interrupts wildly sparking:

```
  631   444 root     R    67392 17.6   0 95.6 ./isr
```

Again on OpenElec 4.1.4 kernel 3.16.1.

**Comment by CurlyMoo** [comment]
I just got confirmation from a XBian user that the problem indeed exists on the XBian staging kernel and in OpenElec but not in Raspbmc:
http://forum.xbian.org/thread-2442-post-24115.html

> Hi, I wanted to moved from the very slow raspbmc to xbian, I have ws2801 LEDs connected to the spi and I get a very weird flickering of random colours on xbian, but this does not happen on raspbmc, this flickering also occurs on openelec.
> [...]
> xbian - 3.15.8
> openelec - 3.16.1
> raspbmc - 3.12.21

(After installing kernel version 3.12.y)

> Great!! booting up fine and hyperion is working flawless, do you think this is the only solution to this problem? or is there going to be an update soon?

**Comment by CurlyMoo** [comment]
Any updates on this?

**Comment by popcornmix** [comment]
I've had a look and can't see a problem with the kernel.

I've put a jumper between GPIO17 and GPIO18

I'm running isr.c with all interrupts except:
  wiringPiISR (1, INT_EDGE_FALLING, &myInterrupt1) ;
commented out. 

```
$ sudo ./isr 
Waiting ... 
```

from another session I do (as root):

```
echo 17 > /sys/class/gpio/export 
echo out > /sys/class/gpio/gpio17/direction
echo 1 > /sys/class/gpio/gpio17/value
echo 0 > /sys/class/gpio/gpio17/value
```

This causes the first session to wake up an claim millions of interrupts have occurred.
I've added logging to kernel and:

```
[ 1075.212163] bcm2708_gpio_irq_mask
[ 1075.212210] bcm2708_gpio_irq_set_type: irq:274 type:1
[ 1075.212222] bcm2708_gpio_irq_unmask
[ 1125.093361] bcm2708_gpio_interrupt
[ 1125.093386] bcm2708_gpio_irq_mask
[ 1125.093411] bcm2708_gpio_irq_unmask
```

Only 1 interrupt has occurred. This is confirmed with:

```
cat /proc/interrupts 
           CPU0       
 16:          0   ARMCTRL  bcm2708_fb dma
 52:          1   ARMCTRL  BCM2708 GPIO catchall handler
 65:          8   ARMCTRL  ARM Mailbox IRQ
 66:          1   ARMCTRL  VCHIQ doorbell
 75:   10457293   ARMCTRL  dwc_otg, dwc_otg_pcd, dwc_otg_hcd:usb1
 77:        480   ARMCTRL  bcm2708_sdhci (dma)
 83:       4625   ARMCTRL  uart-pl011
 84:       2437   ARMCTRL  mmc0
 99:      82263   ARMCTRL  arch_timer
274:          1      GPIO  gpiolib
IPI0:          0  CPU wakeup interrupts
IPI1:          0  Timer broadcast interrupts
IPI2:          0  Rescheduling interrupts
IPI3:          0  Function call interrupts
IPI4:          0  Single function call interrupts
IPI5:          0  CPU stop interrupts
IPI6:          0  IRQ work interrupts
IPI7:          0  completion interrupts
Err:          0
```

isr is stil taking lots of CPU and making the system slow. Kill that and cpu is idle.
If isr is run again, then it is quiet (if interrupts were going wild you may expect them to keep triggering.

So, so far I can't see the problem in the kernel. It may be a problem in wiringPi. 
Possibly https://github.com/WiringPi/WiringPi/blob/master/wiringPi/wiringPi.c#L1033 is not the right way to clear an interrupt.

Firstly - do you see the same behaviour as me?
Make sure the kernel you are using has this patch in:
https://github.com/raspberrypi/linux/commit/93cbf976b7a7d19b0daf5f2ac054aff33425fa74
otherwise you will have gpio related interrupt problems. (recent 3.12 and 3.16 trees will have this - other trees probably won't).

**Comment by CurlyMoo** [comment]
Can you test with https://github.com/pilight/wiringX? It contains the code from wiringPi with some improvements from my side.

**Comment by popcornmix** [comment]
Yes, that's it. Adding:

```
lseek(sysFds[pin], 0, SEEK_SET);
```

to wriringpi's waitForInterrupt now clears the interrupt properly and we just see single interrupts from the isr.c example.

**Comment by CurlyMoo** [comment]
So, XBian dev's did the job ones again :wink: Great that is it working. I will test as soon as possible.

**Comment by CurlyMoo** [comment]
Yep, solved!
